csharp/src/Microsoft.ML.OnnxRuntime/DisposableNamedOnnxValue.cs

```diff
@@ -125,15 +125,108 @@ internal static DisposableNamedOnnxValue CreateFromOnnxValue(string name, IntPtr
+                            return DisposableNamedMapFromNativeMap<Int64, float>(string.Empty, nativeOnnxValueMapKeys, nativeOnnxValueMapValues);
+                        case TensorElementType.String:
+                            return DisposableNamedMapFromNativeMap<string, float>(string.Empty, nativeOnnxValueMapKeys, nativeOnnxValueMapValues);
+                        default:
+                            throw new NotSupportedException("Map of element type: " + elemType + " is not supported");
+                    }
+                default:
+                    throw new NotSupportedException("OnnxValueType : " + onnxValueType + " is not supported");
+            }
+        }
 
-        private static DisposableNamedOnnxValue NameOnnxValueFromNativeTensor<T>(string name, IntPtr nativeOnnxValue)
+        private static DisposableNamedOnnxValue DisposableNamedOnnxValueFromNativeTensor<T>(string name, IntPtr nativeOnnxValue)
         {
             NativeOnnxTensorMemory<T> nativeTensorWrapper = new NativeOnnxTensorMemory<T>(nativeOnnxValue);
             DenseTensor<T> dt = new DenseTensor<T>(nativeTensorWrapper.Memory, nativeTensorWrapper.Dimensions);
             return new DisposableNamedOnnxValue(name, dt, nativeTensorWrapper);
         }
 
+        private static DisposableNamedOnnxValue DisposableNamedMapFromNativeMap<K,V>(string name, IntPtr nativeOnnxValueKeys, IntPtr nativeOnnxValueValues)
+        {
+            var map = new Dictionary<K, V>();
+            IntPtr typeInfo = IntPtr.Zero;
+            IntPtr typeInfoTensor = IntPtr.Zero;
+
+            var onnxValueType = NativeMethods.OrtGetValueType(nativeOnnxValueKeys);
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTypeInfo(nativeOnnxValueKeys, out typeInfo));
+            var meta = InferenceSession.GetMetadataFromTypeInfo(typeInfo);
+
+            var nativeTensorWrapperKeys = new NativeOnnxTensorMemory<K>(nativeOnnxValueKeys).GetSpan();
+            var nativeTensorWrapperValues = new NativeOnnxTensorMemory<V>(nativeOnnxValueValues).GetSpan();
+
+            for (int i=0; i < nativeTensorWrapperKeys.Length; i++)
+            {
+                if (typeof(K) == typeof(string))
+                {
+                    // TODO: add support for string
+                }
+                else
+                {
+                    // types K and V are always valid blittable types
+                    var k = (K) Convert.ChangeType(nativeTensorWrapperKeys[i], typeof(K));
+                    var v = (V) Convert.ChangeType(nativeTensorWrapperValues[i], typeof(V));
+                    map.Add(k, v);
+                }
+            }
 
+            // release native objects
+            if (typeInfo != IntPtr.Zero)
+            {
+                NativeMethods.OrtReleaseTypeInfo(typeInfo);
+            }
+            NativeMethods.OrtReleaseValue(nativeOnnxValueKeys);
+            NativeMethods.OrtReleaseValue(nativeOnnxValueValues);
```

Comment:
If you have created the NativeOnnxTensorWrapper<T> on these memories, then why releasing them? what will back the wrapper then?

Reply:
When a map is constructed, the contents are copied into managed memory. The native data structure cannot be used as a backing store for sequences or maps (only tensors).

Reply:
if you are copying then why do we need to create the MemoryManager<T> (NativeOnnxTensorMemory)? Who is disposing the MemoryManager? if this is just for decoding the Key and Value vectors? in that case we need to factor out the decoding part from the memory management part.

Reply:
The NativeOnnxTensorMemory class is still still used as an accessor class in this particular function, with an extra method added "public Memory<String> MemoryString()" for special case when the string values have to be chunked and decoded, used to construct a dense tensor of **string** type.

 Maps are implemented as two tensors (one for keys, another for values) -- so using NativeOnnxTensorMemory to access the map keys and values simplifies the flow tremendously.

The decoding logic inside NativeOnnxTensorMemory seems like a natural fit -- if we consider  it as an native/managed layer abstraction for tensors (of all primitive types previously, and now string types as well). So, it's supposed to provide Span-like access to native memory of all types. Otherwise, the rest of the classes will need to know how to parse with and work with strings separately, which seems a bit less elegant.

Reply:
seems OK. may be for the sake of consistency, just dispose the 2 NativeOnnxTensorMemory objects instead of directly releasing the native memories here.

Reply:
> seems OK. may be for the sake of consistency, just dispose the 2 NativeOnnxTensorMemory objects instead of directly releasing the native memories here.

Done. See latest version.